crates/sui-core/src/checkpoints/checkpoint_executor/mod.rs

```diff
@@ -609,11 +609,35 @@ async fn execute_transactions(
                     )
                     .collect();
 
+                if missing_digests.is_empty() {
+                    // All effects just become available.
+                    continue;
+                }
+
                 warn!(
-                    "Transaction effects for tx digests {:?} checkpoint not present within {:?}. ",
+                    "Transaction effects for checkpoint tx digests {:?} not present within {:?}. ",
                     missing_digests,
                     log_timeout_sec * periods,
                 );
+
+                // Print out more information for the 1st pending transaction, which should have
+                // all of its input available.
+                let pending_digest = missing_digests.first().unwrap();
```

Comment:
Why only the first? Are we assuming here that there are not more than one non-causally dependent transactions in the checkpoint that still haven't executed?

Reply:
Yes as you mentioned, it is possible for subsequent transactions to also be independent of other transactions in the checkpoint, yet still pending. But it seems more complicated sort them out. For now, just printing one of the stuck transaction should hopefully give us the necessary info for debugging.